cpp/src/arrow/type-test.cc

```diff
@@ -480,6 +482,132 @@ TEST(TestStructType, GetChildIndex) {
   ASSERT_EQ(-1, struct_type.GetChildIndex("not-found"));
 }
 
+TEST(TestDictionaryType, Equals) {
+  auto t1 = dictionary(int8(), ArrayFromJSON(int32(), "[3, 4, 5, 6]"));
+  auto t2 = dictionary(int8(), ArrayFromJSON(int32(), "[3, 4, 5, 6]"));
+  auto t3 = dictionary(int16(), ArrayFromJSON(int32(), "[3, 4, 5, 6]"));
+  auto t4 = dictionary(int8(), ArrayFromJSON(int16(), "[3, 4, 5, 6]"));
+  auto t5 = dictionary(int8(), ArrayFromJSON(int32(), "[3, 4, 7, 6]"));
```

Comment:
I'm quite glad we have this function now =) We might consider adding even more testing sugar so that we can write `jsarray::int64("[3, 4, 5, 6]")` or some such

Reply:
`"[3, 1, 4, 2]"_array(int64())`